packages/eslint-plugin-stories/src/rules/csf-v3-types.ts

```diff
@@ -0,0 +1,160 @@
+          const hasRenderFunctionWithArgument = expression.properties.some(
+            (property) => {
+              // Is this property named "render"?
+              if (
+                property.type === 'Property' &&
+                property.key.type === 'Identifier' &&
+                property.key.name === 'render'
+              ) {
+                // Is the value some sort of function?
+                if (
+                  property.value.type === 'ArrowFunctionExpression' ||
+                  property.value.type === 'FunctionExpression'
+                ) {
+                  // Does this function take any arguments?
+                  return property.value.params.length >= 1;
+                }
+              }
+            },
+          );
+
+          const definitelyNeedsType =
+            isEmptyObject || hasRenderFunctionWithArgument;
+
+          if (definitelyNeedsType) {
+            const typeAnnotation = declarator.id.typeAnnotation;
+
+            if (!typeAnnotation) {
+              context.report({ node, messageId: 'specifyType' });
+            }
+          }
+        }
+      },
+    };
+  },
+};
+
+export default rule;
+
+/**
+ * Component story format (CSF) v3 story.
+ *
+ * @example
+ * export const Primary = {
+ *   render: (args) => <Button {...args} />,
+ * }
+ */
+interface ObjectStory extends TSESTree.ExportNamedDeclaration {
+  declaration: ObjectStoryDeclaration;
+}
+
+/**
+ * The story object being exported.
+ *
+ * @example
+ * const Primary = {
+ *   render: (args) => <Button {...args} />,
+ * }
+ */
+interface ObjectStoryDeclaration extends TSESTree.VariableDeclaration {
+  declarations: [ObjectStoryDeclarator];
+  kind: 'const';
+}
+
+/**
+ * Name and value of the story. Includes TypeScript type, if present.
+ *
+ * @example
+ * Primary = {
+ *   render: (args) => <Button {...args} />,
+ * }
+ *
+ * @example
+ * Primary: StoryObj<Args> = {
+ *   render: (args) => <Button {...args} />,
+ * }
+ */
+interface ObjectStoryDeclarator extends TSESTree.VariableDeclarator {
+  init: TSESTree.ObjectExpression;
+}
+
+/**
+ * Determine if a named export's abstract syntax tree (AST) represents a Story object.
+ */
+// prettier-ignore
+function isObjectStory(node: TSESTree.ExportNamedDeclaration): node is ObjectStory {
+  // Declaration: the thing being exported.
+  // i.e. `const foo = {}`
+  const declaration = node.declaration;
+
+  // Is this thing a "const" variable (not a function or class)?
+  if (declaration?.type === 'VariableDeclaration' && declaration.kind === 'const') {
+    // Declarator: name and value of the variable.
+    // i.e. `foo = {}`
+    const declarator = declaration.declarations[0];
+
+    // The "initial" expression is the value of the variable.
+    // i.e.
+    const expression = declarator.init;
+
+    // If the value is an object literal, assume its a story.
+    return expression?.type === 'ObjectExpression';
+  }
+  return false;
```

Comment:
This whole function exists just to cleanup the rule definition.

But I'm not going to pretend like it's the prettiest code I ever wrote. Let me know if you see ways to clean it up, or if it's not worth extracting in the first place.